### ISSUE TITLE
Return values by reference

### ISSUE THREAD

**Issue by Rogiel** [skypjack/meta#2]
First of all: great work! The interface provided by this library is amazing! I am considering integrating it into my toy c++ game engine to provide reflection for serialization of the engine components.

However, it seems that functions that return by-reference are not supported and will always return a copy. See the following snippet:

```c++
    struct MyValue {
        int value = 0;
        int& get() {
            return value;
        }
    };

    meta::reflect<MyValue>("MyValue")
            .func<&MyValue::get>("get");

    MyValue value;
    meta::any i = meta::resolve("MyValue")
            .func("get")
            .invoke(value);
    value.value = 1;
    std::cout << i.cast<int>() << std::endl;
```

The output would be expected to be `1`. This code, however, outputs `0`. This both unintuitive and significantly limits the possibilities of the library:

 - you can't return non-copyable types by reference
 - you can't mutate any returned member

**Comment by skypjack** [comment]
It's due to `meta::any` being an opaque container that literally copies things inside it to erase their types.  
You should use pointers or `std::reference` for you purposes. You can use external functions as fake members and set them as getters/setter for a property if you don't want to modify your types.

I'm not that sure there exists a way to push around actual references to be honest.  
Did you look at the codebase and have something to suggest?

**Comment by Rogiel** [comment]
> 
> 
> It's due to `meta::any` being an opaque container that literally copies things inside it to erase their types.
> You should use pointers or `std::reference` for you purposes. You can use external functions as fake members and set them as getters/setter for a property if you don't want to modify your types.
> 
> I'm not that sure there exists a way to push around actual references to be honest.
> Did you look at the codebase and have something to suggest?

Yeah. I took a look at how rttr does it and they have a policy tag you can add to the bindings:

```c++
.method("test", &A::test)(
    policy::meth::return_ref_as_ptr
)
```

They use the same syntax to add metadata objects (properties in `meta`), which is pretty nice.

Overall I can see a few possible solutions to this:
 1. have a variation of `meta::func::invoke()` that returns a `meta::handle`. This could get weird and unsafe when returning by value.
 2. Have `meta::any` support references types.
 3. Implicitly convert T& into T* (or `std::reference_wrapper`) when assigning to `meta::any`.

Maybe couple the 3rd approach with something like `rttr` does, by adding a tag or a policy when declaring the `func()`. Personally I would prefer the 2nd approach, but this might be the same discussion as with `std::optional<T&>`.

**Comment by skypjack** [comment]
I think the sole way to get 2 working is by referring to references by means of pointers.  
You cannot really copy or assign a reference, because assigning to a reference goes down to assigning to the object it refers to.  
Therefore `any` won't work any time soon with pure references.

Probably a mix of 2 and 3 could work, that is to allow `any` to work transparently with references without any need for the user to explicitly ask for it.  
Conversion would still be valid too. A reference to a reference collapse in a reference after all, so it would make sense.

What about?

**Comment by Rogiel** [comment]
Yeah. You can't reassign a T&, but you can store it as a T* and make it return `type` that is a reference. For example, `meta::type::is_reference()` (this doesn't exist right now, but it could be added) should return `true`, thus being a different type. It could also allow `data` members to be returned as references, but I don't know if this is interesting or not. A lot of this implementation could be shared for T* and T&, since the implementation will be very similar to T* but I think it is conceptually different.

To avoid unintended construction of `any`s with a reference we could do some tag dispatch using a `any_ref` tag, like this:

```c++
struct any::ref_tag {};

// two new constructors
any::any(any::ref_tag, const Type& type); // type is reference and const
any::any(any::ref_tag, Type& type); // type is reference and non-const

// a helper static function
static any any::ref(Type& type) { // and cref that takes const Type&
    return any(ref_tag{}, type);
}

int myInt;
any intRef = any::ref(myInt); // bound to int&
any intConstRef = any::cref(myInt); // bound to const int&

intRef.cast<int*>(); // should fail
intRef.cast<int>(); // should fail? (idk)
intRef.cast<int&>(); // should work

intRef.convert<int>(); // should work if copyable. Fails if not.
intRef.convert<int*>(); // should work? (maybe, idk)
```

In short, `any` should behave like a `std::reference_wrapper` when it is holding a reference. The problem with wrapping values in a `std::reference_wrapper` is that you cannot pass it as an argument to another function later, since it is not of type `T`, but `std::reference_wrapper<T>` which are totally unrelated types.

**Comment by skypjack** [comment]
The problem with returning the meta type for `T &` when you work with a reference is that it doesn't contain meta data and members, because you registered them with `T` and not with `T &`.

**Comment by Rogiel** [comment]
> 
> 
> The problem with returning the meta type for `T &` when you work with a reference is that it doesn't contain meta data and members, because you registered them with `T` and not with `T &`.

Isn't this already the case when wrapping a `T*`?

Maybe when iterating methods, data, etc it should always decay into `T` (for iteration only). `clazz` should always return `T` (remove const, & and * -- everything) and `self` should be what `clazz` currently is since it is a more proper name to what is does. We would also need to `normalize` between references, pointers and values before forwarding them to `invoke`.

 - During iteration we should iterate on `node.clazz()` rather than `node` then the behavior is consistent across all types.
 - Since different types might have different storage semantics on `any`, before `invoke` arguments should go through `normalize` to have their values normalized to a `T*` and then `internal::invoke` can do it's thing.

**Comment by skypjack** [comment]
I agree. Probably the error is that of trying to stick with the type system of the language.  
When it comes to working with the reflection system, everything should be an opaque reference to an actual value, no matter if the underlying property is a pointer or a reference or whatever.

**Comment by Rogiel** [comment]
> 
> 
> I agree. Probably the error is that of trying to stick with the type system of the language.
> When it comes to working with the reflection system, everything should be an opaque reference to an actual value, no matter if the underlying property is a pointer or a reference or whatever.

I think I can get some proof of concept working tonight, will open another PR with that implemented, since this is orthogonal to references.

**Comment by skypjack** [comment]
It's quite a big change actually. Moreover, I'm out of home until Sunday this week but I want to think if completely erasing the underlying type system is the right thing to do.  
It would give us a _reflection language_ that abstracts completely from the tool but it has also drawbacks. As an example, you won't be able anymore to get a `T**` out from it if everything is treated opaquely and returned through references (mainly because you won't be able anymore to get variables of type `T*`).

**Comment by Rogiel** [comment]
> 
> 
> It's quite a big change actually. Moreover, I'm out of home until Sunday this week but I want to think if completely erasing the underlying type system is the right thing to do.
> It would give us a _reflection language_ that abstracts completely from the tool but it has also drawbacks. As an example, you won't be able anymore to get a `T**` out from it if everything is treated opaquely and returned through references (mainly because you won't be able anymore to get variables of type `T*`).

You still can get a `T*` or `T**`.

```C++
T* ptr = ...;
meta::any any = ptr;
any.type() == meta::resolve<T*>();
meta::func f = any.type().func("abc"); // returns func from T this basically becomes as-if any.type().remove_pointer().func("abc");
f.invoke(any); // is valid now as it should be for T and T&
```

**Comment by skypjack** [comment]
Hi @Rogiel I'm a bit late but these have been busy weeks. Sorry.  
I thought a lot about this and took my time to look into other implementation of reflection or reflection-like systems (eg Qt meta system or rttr).  
To be honest, I don't think an opaque conversion is the best choice, because it forbids you from going the other way around. Simply put, if you want to get the pointer instead of a reference to the pointed object, you now have the same problem if `any` silently returns you references in all cases.

Qt meta system does something similar to `meta` actually, it copies everything back and forth using `QVariant` and all what you can do is to get, modify, then set the property as a whole to update it.  
On the other hand, rttr offers you an extra parameter to use when you bind a property so s to say - _treat references as actual references_.  
I'd rather try to experiment with something along this line. It shouldn't be much difficult to modify `any` so as to deal with reference wrappers in this case. It seems to me the more flexible solution.  
What do you think about?

--- 

I'll try to keep up with this issue on a daily basis now. Sorry if it took me a while to reply so far.

**Comment by Rogiel** [comment]
> 
> 
> Hi @Rogiel I'm a bit late but these have been busy weeks. Sorry.
> I thought a lot about this and took my time to look into other implementation of reflection or reflection-like systems (eg Qt meta system or rttr).
> To be honest, I don't think an opaque conversion is the best choice, because it forbids you from going the other way around. Simply put, if you want to get the pointer instead of a reference to the pointed object, you now have the same problem if `any` silently returns you references in all cases.

I am not sure I understand the problem here. Can you exemplify?

>
> Qt meta system does something similar to `meta` actually, it copies everything back and forth using `QVariant` and all what you can do is to get, modify, then set the property as a whole to update it.
> On the other hand, rttr offers you an extra parameter to use when you bind a property so s to say - _treat references as actual references_.
> I'd rather try to experiment with something along this line. It shouldn't be much difficult to modify `any` so as to deal with reference wrappers in this case. It seems to me the more flexible solution.
> What do you think about?
> 
> I'll try to keep up with this issue on a daily basis now. Sorry if it took me a while to reply so far.

I think this is already implemented this way, if I understood you correctly. You can't create a `any` holding a reference without using `any::ref/cref`, it defaults to passing by value (which is the less error prone method). If you want (and need) a reference, you forward it using `any::ref`. This is what #3 implements.

You can definitely differentiate between a `any` with a value, a pointer or a reference. Their `any::type()` return different values, but you should also be able to convert between them. I will use a template-like notation here to better exemplify what a given `any` is holding, this is not actual code.

You can convert between references and values, and they behave just as they would in C++:
- if `T` is copyable: `any<T&>::convert<T>()` and `any<T*>::convert<T>()` returns a `any<T>`  (i.e. it copies the object)
- `any<T>::convert<T&>()` returns a `any<T&>` that points to the value stored in the `any<T>`, as-if `T v; T& r = v;`.
- `any<T>::convert<T*>()` returns a `any<T*>` that points to the value stored in the `any<T>`, as-if `&value`.

You can still extract the original type stored by a `any`:
- `any<T>::type() == meta::resolve<T>()`
- `any<T&>::type() == meta::resolve<T&>()`
- `any<T*>::type() == meta::resolve<T*>()`
- `any<T>::type() != any<T&>::type() != any<T*>::type()`

Given a `void f(T&)`, you should be able to invoke as:
- `func::invoke(any<T>)`
- `func::invoke(any<T&>)`
- `func::invoke(any<T*>)` (see #4)

We could still offer some helper methods that allow better working with abstract `any` types like: `dereference`, `make_reference` and `make_pointer` that calls the respective `convert()` with a erased type. This is mostly for convenience and allows you to convert a reference into a value (or the other way around) without knowing it's concrete type.

This type of implicit conversion is what #4 will allow (currently only for pointers):

```
void f1(T a);
void f2(T* a);

meta::func<f1>::invoke(any<T>); // copies
meta::func<f2>::invoke(any<T>); // uses &value

meta::func<f1>::invoke(any<T*>); // dereferences AND copies
meta::func<f2>::invoke(any<T*>); // just passes the pointer
```

After #4, #3 can implement a similar mechanism with references:

```
void f1(T a);
void f2(T& a);

meta::func<f1>::invoke(any<T>); // copies
meta::func<f2>::invoke(any<T>); // takes a reference and gives it to f2

meta::func<f1>::invoke(any<T&>); // the any is holding a reference, but it will be copied when being invoked as-if f1(T{})
meta::func<f2>::invoke(any<T&>); // just passes the reference
```

About these last two examples, I don't think the user should really care if the function is taking it's argument by value, by reference or as a pointer. As long as I can can convert the types between them, I simply what to invoke the function and have that "thing" inside `any` passed to it.

**Comment by skypjack** [comment]
Got it. What about #3 instead? I mean, it doesn't compile because it's not backward compatible. Do you plan to add something more to the PR or can I use it as a base along this road by reworking the proposal to an extent?

**Comment by Rogiel** [comment]
> 
> 
> Got it. What about #3 instead? I mean, it doesn't compile because it's not backward compatible. Do you plan to add something more to the PR or can I use it as a base along this road by reworking the proposal to an extent?

It compiles fine. It's missing some tests (code coverage dropped and coveralls rejected it because of that). But it breaks some backwards compatibility because reference and const information is now kept on the type. You can what changed/broke by looking at the changes made to the tests.

**Comment by skypjack** [comment]
I'm closing this because the PRs to which it refers are no longer valid and have sever conflicts with `master`. I'm interested in the proposed changes, but I cannot merge them unfortunately. Feel free to reopen it if you're interested to contribute and review the proposal. Thank you.
